src/runtime_src/core/common/usage_metrics.cpp

```diff
@@ -0,0 +1,516 @@
+      kernel_tree.put("avg_run_time", std::to_string(avg_run_time) + " us");
+    }
+    kernel_array.push_back(std::make_pair("", kernel_tree));
+  }
+
+  return kernel_array; 
+}
+
+static bpt::ptree
+get_hw_ctx_ptree(const std::vector<hw_ctx_metrics>& hw_ctx_vec)
+{
+  bpt::ptree hw_ctx_array;
+
+  uint32_t ctx_count = 0;
+  for (const auto& ctx : hw_ctx_vec) {
+    bpt::ptree hw_ctx;
+    hw_ctx.put("id", std::to_string(ctx_count));
+    hw_ctx.put("xclbin_uuid", ctx.xclbin_uuid.to_string());
+
+    // add buffer info
+    hw_ctx.add_child("bos", get_bos_ptree(ctx.bos_met));
+
+    // add kernel info
+    hw_ctx.add_child("kernels", get_kernels_ptree(ctx.kernel_metrics_vec));
+
+    hw_ctx_array.push_back(std::make_pair("", hw_ctx));
+    ctx_count++;
+  }
+
+  return hw_ctx_array;
+}
+
+static void
+print_usage_metrics()
+{
+  bpt::ptree thread_array;
+
+  uint32_t t_count = 0;
+  // iterate over all threads
+  for (const auto& [thread_id, dev_metrics_map] : *usage_metrics_map) {
+    bpt::ptree dev_array;
+    // iterate over all devices
+    for (const auto& [dev_id, dev_metrics] : dev_metrics_map) {
+      bpt::ptree dev;
+      dev.put("device_index", std::to_string(dev_id));
+      dev.put("bdf", dev_metrics.bdf);
+      dev.put("bos_peak_count", std::to_string(dev_metrics.bo_peak_count));
+
+      // add global bos
+      dev.add_child("global_bos", get_bos_ptree(dev_metrics.global_bos_met));
+
+      // add hw ctx info
+      dev.add_child("hw_context", get_hw_ctx_ptree(dev_metrics.hw_ctx_vec));
+
+      dev_array.push_back(std::make_pair("device", dev));
+    }
+
+    thread_array.add_child("thread " + std::to_string(t_count), dev_array);
+    t_count++;
+  }
+
+  print_json(thread_array);
+}
+
+// class usage_metrics_logger - class for logging usage metrics
+//
+// Logging objects are created per thread 
+//
+// This class collects metrics from all threads using XRT
+// The metrics are collected in a thread safe manner.
+class usage_metrics_logger : public xrt_core::usage_metrics::base_logger
+{
+public:
+  usage_metrics_logger();
+
+  ~usage_metrics_logger();
+
+  void
+  log_device_info(const xrt_core::device*) override;
+
+  void 
+  log_hw_ctx_info(const xrt::hw_context_impl*) override;
+
+  void 
+  log_buffer_info_construct(device_id, size_t, const xrt_core::hwctx_handle*) override;
+
+  void 
+  log_buffer_info_destruct(device_id) override;
+
+  virtual void
+  log_buffer_sync(device_id, const xrt_core::hwctx_handle*, size_t, xclBOSyncDirection) override;
+
+  void
+  log_kernel_info(const xrt_core::device*, const xrt::hw_context&, const std::string&, size_t) override;
+
+  void
+  log_kernel_run_info(const xrt::kernel_impl*, const xrt::run_impl*, ert_cmd_state) override;
+
+private:
+  device_metrics_map m_dev_map;
+  std::shared_ptr<metrics_map> map_ptr;
+};
+
+usage_metrics_logger::
+usage_metrics_logger() : map_ptr(usage_metrics_map)
+{
+  thread_count++;
+}
+
+usage_metrics_logger::
+~usage_metrics_logger()
```

Comment:
We need try / catch to resolve 
- https://scan9.scan.coverity.com/reports.htm#v36614/p13446/fileInstanceId=123592600&defectInstanceId=9463167&mergedDefectId=330809
- https://scan9.scan.coverity.com/reports.htm#v36614/p13446/fileInstanceId=123592600&defectInstanceId=9463171&mergedDefectId=330815
- https://scan9.scan.coverity.com/reports.htm#v36614/p13446/fileInstanceId=123592600&defectInstanceId=9463146&mergedDefectId=330816